operator/rpc_client.go

```diff
@@ -42,82 +75,172 @@ func (c *AggregatorRpcClient) dialAggregatorRpcClient() error {
 	return nil
 }
 
-func (c *AggregatorRpcClient) sendRequest(sendCb func() error, maxRetries int, retryInterval time.Duration) {
-	if c.rpcClient == nil {
-		c.logger.Info("rpc client is nil. Dialing aggregator rpc client")
-		err := c.dialAggregatorRpcClient()
+func (c *AggregatorRpcClient) InitializeClientIfNotExist() error {
+	c.rpcClientMutex.Lock()
+	defer c.rpcClientMutex.Unlock()
+
+	if c.rpcClient != nil {
+		return nil
+	}
+
+	return c.dialAggregatorRpcClient()
+}
+func (c *AggregatorRpcClient) onTick() {
+	tickerC := c.resendTicker.C
+	for {
+		// TODO(edwin): handle closed chan
+		<-tickerC
+
+		// Critically ugly section
+		{
+			c.unsentMessagesMutex.Lock()
+			if len(c.unsentMessages) == 0 {
+				c.unsentMessagesMutex.Unlock()
+				continue
+			}
+			c.unsentMessagesMutex.Unlock()
+		}
+
+		err := c.InitializeClientIfNotExist()
 		if err != nil {
-			c.logger.Error("Could not dial aggregator rpc client. Not sending signed task response header to aggregator. Is aggregator running?", "err", err)
-			return
+			continue
 		}
+
+		c.tryResendFromDeque()
+	}
+}
+
+// Expected to be called with initialized client.
+func (c *AggregatorRpcClient) tryResendFromDeque() {
+	c.unsentMessagesMutex.Lock()
+	defer c.unsentMessagesMutex.Unlock()
+
+	if len(c.unsentMessages) != 0 {
+		c.logger.Info("Resending messages from queue")
 	}
 
-	c.logger.Info("Sending data to aggregator")
-	for i := 0; i < maxRetries; i++ {
-		err := sendCb()
+	errorPos := 0
+	for i := 0; i < len(c.unsentMessages); i++ {
+		message := c.unsentMessages[i]
+
+		// Assumes client exists
+		var err error
+		var reply bool
+
+		switch message.MessageType {
+		case CheckpointType:
+			signedCheckpointTaskResponse := message.Message.(*coretypes.SignedCheckpointTaskResponse)
+			// TODO(edwin): handle error
+			err = c.rpcClient.Call("Aggregator.ProcessSignedCheckpointTaskResponse", signedCheckpointTaskResponse, &reply)
 
-		if err == nil {
-			c.logger.Info("Data successfully sent to aggregator")
-			return
+		case StateRootType:
+			signedStateRootUpdateMessage := message.Message.(*coretypes.SignedStateRootUpdateMessage)
+			err = c.rpcClient.Call("Aggregator.ProcessSignedStateRootUpdateMessage", signedStateRootUpdateMessage, &reply)
+
+		case OperatorSetUpdateType:
```

Comment:
I didn't like the idea of passing raw interface into db methods so I done it this way. Don't have a strong preference tho

Reply:
Sure, not a problem. What I mean is you're using a `MessageType` thing to do checks and then casting the raw interface. For example, instead of:
```go
		switch message.MessageType {
		case CheckpointType:
			signedCheckpointTaskResponse := message.Message.(*coretypes.SignedCheckpointTaskResponse)
			err = c.rpcClient.Call("Aggregator.ProcessSignedCheckpointTaskResponse", signedCheckpointTaskResponse, &reply)

		case StateRootType:
			signedStateRootUpdateMessage := message.Message.(*coretypes.SignedStateRootUpdateMessage)
			err = c.rpcClient.Call("Aggregator.ProcessSignedStateRootUpdateMessage", signedStateRootUpdateMessage, &reply)

		case OperatorSetUpdateType:
			signedOperatorSetUpdateMessage := message.Message.(*coretypes.SignedOperatorSetUpdateMessage)
			err = c.rpcClient.Call("Aggregator.ProcessSignedOperatorSetUpdateMessage", signedOperatorSetUpdateMessage, &reply)
}
```
You could do:
```go
switch msg := message.Message.(type) {
case *coretypes.SignedCheckpointTaskResponse:
    err = c.rpcClient.Call("Aggregator.ProcessSignedCheckpointTaskResponse", msg, &reply)
case *coretypes.SignedStateRootUpdateMessage:
    err = c.rpcClient.Call("Aggregator.ProcessSignedStateRootUpdateMessage", msg, &reply)
case *coretypes.SignedOperatorSetUpdateMessage:
    err = c.rpcClient.Call("Aggregator.ProcessSignedOperatorSetUpdateMessage", msg, &reply)
}
```

Just seems a lot better to me, and then we could drop the message type field and also the `RpcMessageType`.

Reply:
Done